File: python-solutions/misc/permutations.py
def permute(s):
    perm_rec("", s)


def perm_rec(perm, rest):
    if len(rest) == 0:
        print(perm)
    else:
        for i in range(0, len(rest)):
            ch = rest[i]
            new_rest = rest[0:i] + rest[i+1:]
            perm_rec(perm + ch, new_rest)


permute("abc")

File: rust-solutions/misc/src/bin/permutations.rs
/// Generate (print) the permutations of a string

fn main() {
    permute("abc");
}

fn permute(s: &str) {
    perm_rec(String::new(), s.to_string())
}

fn perm_rec(perm: String, rest: String) {
    match rest.len() {
        0 => println!("{}", &perm),
        _ => {
            for i in 0..rest.len() {
                let c = rest.chars().nth(i).unwrap();
                // build the rest string without the char at position i
                let new_rest = format!("{}{}", &rest[0..i], &rest[i+1..]);

                perm_rec(perm.clone() + &c.to_string(), new_rest);
            }
        }
    }
}